### ISSUE TITLE
[feature] Add support for event notifications

### ISSUE THREAD

**Issue by corollari** [CityOfZion/neon-js#511]
## Motivation
When building interfaces for smart-contract-based applications it's pretty common to want it to react to changes in the contract state in order to make the UI accurately represent the current state. Examples of this could be an exchange that wants to remove an outstanding buy order from the interface when such order is taken or withdrawn, or a game that would display the moves of an opponent the moment they happen.

Right now this can only be achieved through constant querying of the RPC nodes or by creating a custom solution, but something that would serve this need much better would be a system that would allow a developer to subscribe to a smart contract and be notified of event triggers on it, essentially a publish-subscribe system.

As a data point, on Ethereum, a system like this is integrated into all the major libraries and used by almost all dApps. Note that I'm not advocating for this library to become a copy of web3, but pointing out that in a similar network this kind of feature has seen a lot of usage.

## Feature
Integrate a publish-subscribe framework that enables developers to subscribe to a smart contract in order to receive events triggered on it, an example of such api could be:
```
import { api } from "@cityofzion/neon-js";

api.pubsub.subscribe(address, (event)=>doSomething(event));
```

where the second argument is a callback function that would be called every time an event in that contract is triggered.

This is just an idea to illustrate the concept tho, the actual implementation can follow other designs.

## Infrastructure
Unless we turn neon-js clients into nodes capable of interacting with the p2p network and executing scripts on the NeoVM, integrating this feature into neon-js will require some extra infrastructure that collects events being triggered inside smart contracts and broadcasts them to subscribed clients, essentially the publisher of pubsub architecture.

I've written, maintain and run [a server that could serve that purpose](https://github.com/corollari/neo-PubSub). In it's current form it broadcasts events triggered on all neo smart contracts and messages received through the p2p network. Furthermore, the architecture it uses consists of a single neo node that collects events and messages and broadcasts these to a dynamic set of servers running on heroku (this set can scale with demand) which maintain websocket connections with all the subscribers and rebroadcast everything to them, resulting in a highly scalable system, which should be able to support heavy loads.

The system could be made more scalable by using AWS IoT to relay the messages, but I don't think it will reach a load where that is necessary, and if we do then we can migrate it over there.

- Source code: https://github.com/corollari/neo-PubSub/
- Showcase page: https://corollari.github.io/neo-PubSub/

## Extra
Something that might also be interesting to implement is some mechanism that allows a developer to know when a transaction has been finalised, that is, has been included into a block.

**Comment by snowypowers** [comment]
Sounds like a good feature to pickup, i dont imagine there is much work behind this as its just implementing a events handler. Not sure if there is any extra support required for websockets.

Notes:
1. This sounds like a possible plugin on the c# side. Do we have a standard on that side? If it is possible to rope in and get a consensus on a common API interface, it would be easier to buy in.

2. I noticed that the endpoints were designed as `domain/?channel=type`. I would like to know if there is any reasoning behind this API design. Conventional API design would put the resource names as part of the url (`domain/channeltype`) and leave the query params as additional filtering syntax. This will conform nicely to many API tooling and allow us to strongly type each endpoint.

3. Are the payload schemas from c# side? There are some inconsistencies across the payloads (like txid vs txID). Would love to clean it up if possible.

**Comment by corollari** [comment]
Yes, implementation should be pretty simple as it's just a wrapper over websockets. I can do it myself (will send a PR in the following days).

> This sounds like a possible plugin on the c# side. Do we have a standard on that side? If it is possible to rope in and get a consensus on a common API interface, it would be easier to buy in.

I don't think I really understood what you are referring to, but, if you're talking about the source of the events, it is indeed a plugin for the C# client which obtains the events as `NotifyEventArgs.State.ToParameter().ToJson().ToString()` and then sends them to a redis queue from which they are picked up and broadcasted.

The source code for the plugin I'm currently using is https://github.com/safudex/smartbnb/tree/master/NeoPubSub, which is a fork of https://github.com/CityOfZion/neo-plugins-coz/tree/master/NeoPubSub and shares the same interface.

Regarding the interface itself, at the moment we are just broadcasting the internal representation that neo-cli uses, for which there's no standard for. Nevertheless it's been quite stable so far and if there's changes to it in the future I believe that we will be able to adapt them to the current interface.

> endpoints

There's no reasoning behind it, I just followed the convention used by the tool that I forked. I agree with your proposal here and will change the urls to something closer to normal standards.

> Are the payload schemas from c# side? 

Some of them, transaction information comes from the RPC interface and some parts are just set by my code, I will iron these out to have something more consistent.
